text/0161-telemetry-schema-and-resource.md

```diff
@@ -0,0 +1,196 @@
+pull the version number out of the schema url.
+
+### Resource merge logic
+
+Resource merge logic is updated as follows (new mechanisms in bold):
+
+The resulting resource will have the Schema URL calculated as follows:
+
+- If the old resource's Schema URL is empty then the resulting
+  resource's Schema URL will be set to the Schema URL of the updating resource
+- Else if the updating resource's Schema URL is empty then the resulting
+  resource's Schema URL will be set to the Schema URL of the old resource,
+- Else if the Schema URLs of the old and updating resources are the same then
+  that will be the Schema URL of the resulting resource,
+- **Else if the Schema urls of the old and updating resources are compatible**
+  **then the older resource will be converted to newer schema url and merged.**
+- Else this is a merging error (this is the case when the Schema URL of the old
+  and updating resources are not empty and are different).
+
+### Resource conversion logic
+
+SDKs will be expected to provide a mechanism to apply transformations
+(both backwards and forwards) listed in `changes` for any compatible schemas.
+
+**Note: Until conversion logic is available in SDKs, any change to telemetry**
+**schema that relies on `changes` will be considered a breaking, i.e.**
+**they will lead to a major version bump.**
+
+## Trade-offs and mitigations
+
+The major drawback to this proposal is the requirement of making all schema
+migration implementations available in all SDKs.  This is mitigated by the
+following:
+
+- Initially, we could prevent any `rename` alterations to schema for compatible
+  version numbers. `convertToSchema` would check for semver compatibility and
+  just return the original resource if compatible.
+- Alternatively, we can check the newer schema for any transformations in the
+  `all` or `resource` section before failing to convert.
+
+This should allow a slow-rollout of implementation by SDKs by increasing the
+number of successful `convertToSchema` operations they allow.
+
+## Prior art and alternatives
+
+It's a common practice in protocols to negotiate version semantics and for
+newer code to provide backwards-compatible "views" of older APIs for
+compatibility.   This is intended as a stop-gap to prevent immediate breakages
+as the ecosystem of components evolves to the new protocol on independent
+timelines.  Specifically "core" components can update without forcing all
+downstream to components to also update.
+
+### Alternatives
+
+One alternative we can take here is to force all exporters to always be on the
+same version of telemetry schema as the SDK and find a mechanism to warn/prevent
+older exporters from using a newer SDK.
+
+Another alternative is for SDKs to continue propagating resources as-is and
+have exporters issue errors when resource versions don't align with
+expectations.
+
+## Open questions
+
+- What kinds of operations on Resource should be considered compatible?
```

Comment:
This is defined by schema OTEP. Conversion between families is incompatible. Ambiguous "renames" are incompatible.